### ISSUE TITLE
Thoughts about Releases and Branches and Processes in a Codepush World

### ISSUE THREAD

**Issue by hawkrives** [StoDevX/AAO-React-Native#1044]
With Codepush, we'll need to figure out a new scheme for doing release version numbers.

I found this on Twitter:

![image 2017-05-08 15 16 45](https://cloud.githubusercontent.com/assets/464441/25823328/5fb3eca0-3401-11e7-9eeb-9767f2c4dfc5.jpg)

> We are doing minor version bumps for native changes and patch [bumps] for JS [changes].

I like that.

Here are some things that I think we need to be able to do:

- We'll need to be able to roll out bug fixes to multiple minor versions at once. We should be able to, say, disable the SIS > Courses tab on v2.2 and v2.3 from codepush.

- The app stores only support updating people to the latest version of the binary, so we only need to worry about that from Codepush.

- Codepush allows you to target a specific binary version with their `-t, --targetBinaryVersion` parameter.

So here's my proposal:

- We will make a _branch_ for each active minor release
    - This allows us to optionally cherry-pick commits from `master`, or write custom patches, as needed, targeting specific binary versions
    - When we release 2.2.0, we'll make a v2.2 branch. 2.3.0 will get its own new branch (v2.3), but 2.2.1 will be done on the v2.2 branch
- We set up Travis to…
    - automatically publish new _untagged_ commits to those branches to the `staging` channel
    - automatically publish new _tagged_ commits to those branches to the `release` channel
    - … with the `--targetBinaryVersion` argument set to restrict the deployment to the appropriate binary version
- For nightly builds, we will:
    - ~~Make Travis deploy `master` every night to the `nightly` codepush channel~~
    - Travis will deploy the tagged commits from `master` as betas, as soon as you tag them, rather than once a day
    - Hockey betas will be set to update from the `nightly` channel
    - After each native release, we will bump the version to a "beta.1" number
        - for instance, after v2.2 is release, we will bump the versions to `v2.3.0-beta.1`
    - When we merge a native dependency update, we will also bump the beta version
        - For instance, when `react-native-maps` updates, we'd go from `v2.3.0-beta.1` to `v2.3.0-beta.2`
    - Travis will only release new Hockey betas when we tag a new beta version
    - Travis will release Codepush bundles targeted to just the latest beta tag to the nightly channel

---

This will allow us to continue working on `master`, as we normally do, but also backport any js-only bug fixes to the live versions, simply by adding git tags.

**Comment by hawkrives** [comment]
Keep in mind that almost all of that proposal would be implemented in travis, automatically.

The biggest differences for us would be in remembering to tag new betas from master.

(and backporting bugfixes, but that’s an entirely new process anyway.)

**Comment by hawkrives** [comment]
I've disabled the nightly cron job for now, at least. Until 2.2 is released, I think we're good to just do tagged betas regardless of if we want to go all the way with this idea.

**Comment by drewvolz** [comment]
This all sounds great to me.

**Comment by hawkrives** [comment]
> @drewvolz: I did have a question, though. You described publishing untagged commits to staging/nightly, but those are two separate channels. Are we reserving staging for commits that have been thru nightly and are verified as good? I get that travis will release codepush bundles to nightly, and new hockey betas, but I am confused what we deem as staging.

Given that `nightly` is pretty much going to be incompatible with the release channels, I don’t think we can promote from nightly to anything else.

So my suggestion is more along the lines of, `nightly` is incompatible with `release`, but we still want to be able to test releases before everyone gets them. So, whenever we commit to a release branch, that commit is sent to the staging channel with an appropriate targetBinaryVersion, so we can work with multiple release channels.

When we get a good commit, we’ll just tag it, so we have a record of what commits were what versions, and Travis will make a new bundle and release it to release. This completely avoids the code-push promote command, which was not intentional, but I don’t think we need it with this framework.

TL;DR: Travis will create and release bundles, and we’ll essentially never deal with the code-push cli directly. Aside from checking the status of things, I think.

**Comment by elijahverdoorn** [comment]
Love it.

**Comment by hawkrives** [comment]
@elijahverdoorn had a good heuristic for when to release new binaries:

> updating data, whatever. Changing UX, we shoud do a binary.
